### PR TITLE
fix: fix arethetypesworking errors in all packages

### PR DESCRIPTION
**Pull request by gobengo**
Motivation:
* https://github.com/web3-storage/w3up/issues/982

Learned While Devving
* #982 indicated 3 errors coming from atty. Here is how each error type was resolved
  1. `ESM (dynamic import only)` - this is more of a warning than an error. It's warning, rightly, that versions of node that only use CommonJS requires (e.g. `require("@web3-storage/did-mailto")`) won't be able to load these libraries. i.e. these libraries are "ESM-only" in that they only work on runtimes that can do ESM-style `import()`. By default, attw warns on this, but this PR configures attw to ignore instead (via .attw.json https://github.com/web3-storage/w3up/pull/1004/files#diff-a3be4137698f4bdc5777d3831fac4e67d11c102c5e57dd54906ff333cb551aff ).
  2.  [internal resolution error](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/InternalResolutionError.md)
      * Insight via https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/InternalResolutionError.md#declarations-and-javascript-match-but-are-not-checked-with-correct-settings
        > This problem can occur with tsc alone if the library author compiles with --moduleResolution node (also known as node10) or bundler, especially in combination with a package.json that has "type": "module" 
      * This attw error was resolved by this PR but this PR changing all packages `tsconfig.json` to use moduleResolution=NodeNext (not `Node`). To do that, many type imports had to change so the imported module identifier includes a file path suffix
  3. [no types](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/UntypedResolution.md) - Most of these were due to the relevant package having sub packabe submodule listed in package.json `exports` but without a corresponding js file existing (e.g. it was an old package submodule that has since been removed)

Also
* added an npm script `attw` to run attw for the package. Now can run attw for all packages like `pnpm -r attw` from monorepo root
* gitlab ci workflows for the two packages shown in #982 now run attw

### CODE REVIEW

File: packages/access-client/package.json [hunk 1]
```diff
@@ -24,12 +25,12 @@
     "rc": "npm version prerelease --preid rc"
   },
   "exports": {
-    ".": "./src/index.js",
-    "./agent": "./src/agent.js",
-    "./drivers/*": "./src/drivers/*.js",
-    "./stores/*": "./src/stores/*.js",
-    "./types": "./src/types.js",
-    "./encoding": "./src/encoding.js"
+    ".": "./dist/src/index.js",
```
Comment (gobengo) [inline packages/access-client/package.json:28]: changing where this points to resolved some atty errors. When we do it this way and let typescript also write the type declaration files with all its defaults next to the source files, then when another tsc in a depending lib finds all this config, it'll be able to find the src,types,etc again. I learned some of this from https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseESM.md#common-causes

File: packages/access-client/src/agent-data.js [hunk 2]
```diff
@@ -6,16 +6,16 @@ import { CID } from 'multiformats'
 import { Access } from '@web3-storage/capabilities'
 import { isExpired } from './delegations.js'
 
-/** @typedef {import('./types').AgentDataModel} AgentDataModel */
+/** @typedef {import('./types.js').AgentDataModel} AgentDataModel */
```
Comment (gobengo) [inline packages/access-client/src/agent-data.js:9]: This is one of many small changes to ensure all imported module identifiers (e.g. type imports in comments like this) have a file suffix, which is required by `tsc` now that this PR enables `moduleResolution=NodeNext` for all tsc packages.

File: tsconfig.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
     "allowJs": true,
     "checkJs": true,
     "target": "ES2022",
-    "module": "ES2022",
```
Comment (gobengo) [inline tsconfig.json:9]: This change inspired by
https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/64b1f489736bd48ebb4e3b72fb078b44437d59e0/docs/problems/InternalResolutionError.md
> This problem can occur with tsc alone if the library author compiles with --moduleResolution node (also known as node10) or bundler, especially in combination with a package.json that has "type": "module"

and

> Library authors should always compile/check with --module node16 --moduleResolution node16 or nodenext (note that module im

File: packages/upload-api/src/consumer/has.js [hunk 4]
```diff
@@ -6,6 +6,7 @@ import { Consumer } from '@web3-storage/capabilities'
  * @param {API.ConsumerServiceContext} context
  */
 export const provide = (context) =>
+  // @ts-ignore
```
Comment (travis) [inline packages/upload-api/src/consumer/has.js:9]: could you add a comment here justifying the ignore?

File: packages/upload-client/test/unixfs.test.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import assert from 'assert'
 import { decode, NodeType } from '@ipld/unixfs'
 import { exporter } from 'ipfs-unixfs-exporter'
+// @ts-ignore
```
Comment (travis) [inline packages/upload-client/test/unixfs.test.js:4]: here too plz

**Review by travis — APPROVED** [review]
Looks great! have a couple requests, and I'd feel a lot comfier if @alanshaw or @olizilla got a chance to take a look, but happy to see it merge as-is

File: packages/upload-client/test/unixfs.test.js (hunk 5 above)
Comment (gobengo) [reply packages/upload-client/test/unixfs.test.js]: updated https://github.com/web3-storage/w3up/pull/1004/commits/df5995bbf679029a9a9a4f5c52984a9eacf1c67e

File: packages/upload-api/src/consumer/has.js (hunk 4 above)
Comment (gobengo) [reply packages/upload-api/src/consumer/has.js]: I removed it! good catch. My local checker was in a weird state across my editor and local terminal and ci, but I think it'll work now https://github.com/web3-storage/w3up/pull/1004/commits/e9882a4c4a7507dcf22ea31e090d84dee376b999

File: packages/access-client/package.json (hunk 1 above)
Comment (alanshaw) [reply packages/access-client/package.json:28]: Can you update the npm published files so we don't include src in the root?

**Review by alanshaw — APPROVED** [review]

File: packages/access-client/package.json (hunk 1 above)
Comment (gobengo) [reply packages/access-client/package.json:28]: @alanshaw https://github.com/web3-storage/w3up/pull/1012

**Pull request merged**